src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java

```diff
@@ -0,0 +1,1154 @@
+ *         return newValue.toLowerCase();
+ *     }
+ *
+ *     void read() throws IOException {
+ *         BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
+ *         String s; // violation
+ *         while ((s = reader.readLine()) != null) {
+ *         }
+ *         try (BufferedReader reader1 // ok as 'reader1' is a resource and resources are closed
+ *                                     // at the end of the statement
+ *             = new BufferedReader(new FileReader("abc.txt"))) {
+ *         }
+ *         try {
+ *         } catch (Exception e) {     // ok as e is an exception parameter
+ *         }
+ *     }
+ *
+ *     void loops() {
+ *         int j = 12;
+ *         for (int i = 0; j &lt; 11; i++) { // violation, unused local variable 'i'.
+ *         }
+ *         for (int p = 0; j &lt; 11; p++)   // ok
+ *             p /= 2;
+ *     }
+ *
+ *     void lambdas() {
+ *         Predicate&lt;String&gt; obj = (String str) -&gt; { // ok as 'str' is a lambda parameter
+ *             return true;
+ *         };
+ *         obj.test("test");
+ *     }
+ * }
+ * </pre>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code unused.local.var}
+ * </li>
+ * </ul>
+ *
+ * @since 9.3
+ */
+@FileStatefulCheck
+public class UnusedLocalVariableCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_UNUSED_LOCAL_VARIABLE = "unused.local.var";
+
+    /**
+     * An array of increment and decrement tokens.
+     */
+    private static final int[] INCREMENT_AND_DECREMENT_TOKENS = {
+        TokenTypes.POST_INC,
+        TokenTypes.POST_DEC,
+        TokenTypes.INC,
+        TokenTypes.DEC,
+    };
+
+    /**
+     * An array of scope tokens.
+     */
+    private static final int[] SCOPES = {
+        TokenTypes.SLIST,
+        TokenTypes.LITERAL_FOR,
+        TokenTypes.OBJBLOCK,
+    };
+
+    /**
+     * An array of unacceptable children of ast of type {@link TokenTypes#DOT}.
+     */
+    private static final int[] UNACCEPTABLE_CHILD_OF_DOT = {
+        TokenTypes.DOT,
+        TokenTypes.METHOD_CALL,
+        TokenTypes.LITERAL_NEW,
+        TokenTypes.LITERAL_SUPER,
+        TokenTypes.LITERAL_CLASS,
+        TokenTypes.LITERAL_THIS,
+    };
+
+    /**
+     * An array of unacceptable parent of ast of type {@link TokenTypes#IDENT}.
+     */
+    private static final int[] UNACCEPTABLE_PARENT_OF_IDENT = {
+        TokenTypes.VARIABLE_DEF,
+        TokenTypes.DOT,
+        TokenTypes.LITERAL_NEW,
+        TokenTypes.PATTERN_VARIABLE_DEF,
+        TokenTypes.METHOD_CALL,
+        TokenTypes.TYPE,
+    };
+
+    /**
+     * An array of blocks in which local anon inner classes can exist.
+     */
+    private static final int[] CONTAINERS_FOR_ANON_INNERS = {
```

Comment:
Anonymous inner classes can be declared in compact constructors, why aren't they listed here?

Reply:
I missed records, added support for them now.

Reply:
Resolved.